[WGSL] Implement comparison operator overload and codegen

djg opened the pull request:
#### 3a71603352dcd72cfd85c308e4bea671cff8c56a
<pre>
[WGSL] Implement comparison operator overload and codegen
<a href="https://bugs.webkit.org/show_bug.cgi?id=255321">https://bugs.webkit.org/show_bug.cgi?id=255321</a>
rdar://107923257

Reviewed by Tadeu Zagallo.

Implement comparison operator overload and codegen for WGSL.

This change was complicated by needing to teach the overload matching generator
to handle `vecN&lt;bool&gt;`. Adding a constraint type `Boolean` with one valid type
`Bool` didn&apos;t work when using `Vector[S &lt; Boolean, N]` as the return type of an
operator.

To accept `Vector[Bool, N]` aka `vecN&lt;bool&gt;`, A new type `AbstractScalarType` is
introduced that can hold a concrete type or a type variable. `materialize`,
`calculateRank` and `unify` are updated to handle `AbstractType`,
`AbstractScalarType`, and `AbstractValue`. The &quot;hack&quot; in the generator that
selects between `ParameterizedAbstractType` or `Constructor` is extended to
selected `ParameterizedAbstractType` when any argument is a `Variable` to be
able to handle a vector of a concrete type but with a variable number of
elements.

For completeness, `AbstractMatrix` element was also updated.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/Overload.cpp:
(WGSL::OverloadResolver::materialize const):
(WGSL::OverloadResolver::calculateRank):
(WGSL::OverloadResolver::unify):
(WTF::printInternal):
* Source/WebGPU/WGSL/Overload.h:
* Source/WebGPU/WGSL/TypeDeclarations.rb:
* Source/WebGPU/WGSL/generator/main.rb:
* Source/WebGPU/WGSL/tests/valid/overload.wgsl:

Canonical link: <a href="https://commits.webkit.org/262903@main">https://commits.webkit.org/262903@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff8e2f17753659de54f0503e36c9beecea253c35

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2833 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2902 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2992 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4245 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3268 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2983 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2941 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2490 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2861 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3263 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2528 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4028 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/758 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2511 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2380 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2518 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2562 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3771 "261 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2914 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2324 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2563 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2514 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/732 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2554 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2734 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebGPU/WGSL/Overload.h:
```diff
@@ -74,13 +74,18 @@ using AbstractType = std::variant<
+
 struct AbstractVector {
-    TypeVariable element;
+    AbstractScalarType element;
```
Why not use `AbstractType`? I know it's more permissive, but in practice I don't think it matters, and might even be more correct.

djg replied:
I didn't use `AbstractType` because it leads to a circular use of `AbstractVector`. `AbstractType` is `std::variant<..., AbstractVector, ...>` and when trying to instantiate the `variant` template for the `element` member the compiler has only seen an incomplete type at that point.

```
In file included from /Volumes/WebKit/OpenSource/Source/WebGPU/WGSL/TypeCheck.cpp:26:
In file included from /Volumes/WebKit/OpenSource/Source/WebGPU/WGSL/config.h:26:
In file included from /Volumes/WebKit/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/Assertions.h:61:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.Internal.sdk/usr/include/c++/v1/type_traits:462:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.Internal.sdk/usr/include/c++/v1/__type_traits/is_copy_constructible.h:27:11: error: incomplete type 'WGSL::AbstractVector' used in type trait expression
          __is_constructible(_Tp, __add_lvalue_reference_t<typename add_const<_Tp>::type>)> {};
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.Internal.sdk/usr/include/c++/v1/__type_traits/is_copy_constructible.h:31:49: note: in instantiation of template class 'std::is_copy_constructible<WGSL::AbstractVector>' requested here
inline constexpr bool is_copy_constructible_v = is_copy_constructible<_Tp>::value;
```

tadeuzagallo reviewed (commented):
Overall looks good, but I think it can be simplified.

djg committed:
[WGSL] Implement comparison operator overload and codegen
https://bugs.webkit.org/show_bug.cgi?id=255321
rdar://107923257

Reviewed by Tadeu Zagallo.

Implement comparison operator overload and codegen for WGSL.

This change was complicated by needing to teach the overload matching generator
to handle `vecN<bool>`. Adding a constraint type `Boolean` with one valid type
`Bool` didn't work when using `Vector[S < Boolean, N]` as the return type of an
operator.

To accept `Vector[Bool, N]` aka `vecN<bool>`, A new type `AbstractScalarType` is
introduced that can hold a concrete type or a type variable. `materialize`,
`calculateRank` and `unify` are updated to handle `AbstractType`,
`AbstractScalarType`, and `AbstractValue`. The "hack" in the generator that
selects between `ParameterizedAbstractType` or `Constructor` is extended to
selected `ParameterizedAbstractType` when any argument is a `Variable` to be
able to handle a vector of a concrete type but with a variable number of
elements.

For completeness, `AbstractMatrix` element was also updated.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/Overload.cpp:
(WGSL::OverloadResolver::materialize const):
(WGSL::OverloadResolver::calculateRank):
(WGSL::OverloadResolver::unify):
(WTF::printInternal):
* Source/WebGPU/WGSL/Overload.h:
* Source/WebGPU/WGSL/TypeDeclarations.rb:
* Source/WebGPU/WGSL/generator/main.rb:
* Source/WebGPU/WGSL/tests/valid/overload.wgsl:

Canonical link: https://commits.webkit.org/262903@main
